File: Python/0509-fibonacci-number.py
class Solution:
    def fib(self, n: int) -> int:
        dp = [0] * (n + 1)
        return self.rec(n, dp)

    def rec(self, n: int, dp: List[int]):
        if n == 0 :
            return 0
        elif n == 1 :
            return 1
        elif dp[n - 1] == 0 :
            dp[n - 1] = self.rec(n - 1, dp) + self.rec(n - 2, dp)
            
        return dp[n - 1]

File: Java/0509-fibonacci-number.java
class Solution {
    public int fib(int n) {
        int[] mem = new int[n];
        return rec(n, mem);
    }
    private int rec(int n, int[] mem) {
        if (n <= 0) {
            return 0;
        }
        else if (n <= 2) {
            return 1;
        }
        else if (mem[n - 1] == 0) {
            mem[n - 1] = rec(n - 2, mem) + rec(n - 1, mem);
        }
        return mem[n - 1];
    }
}